File: MCRunner/Instruments/PlayableBars.cs
using System;
using System.Collections.Generic;

namespace MCRunner.Instruments
{
    public class PlayableBars : Bars
    {
        private IEnumerable<Bar> BarsToPlay { get; set; }

        public PlayableBars(IEnumerable<Bar> barsToPlay)
        {
            BarsToPlay = new List<Bar>(barsToPlay);
        }

        public void Play(Action action)
        {
            Play(bar => action());
        }

        public void Play(Action<Bar> action)
        {
            bars.Clear();

            foreach (var bar in BarsToPlay)
            {
                bars = bars.Add(bar);
                ReloadBarSeries();
                action(bar);
            }

            BarsToPlay = null;
        }
    }
}

File: python/mcrunner/instruments/playable_bars.py
from typing import Iterable, Callable
from .bars import Bars
from .bar import Bar

class PlayableBars(Bars):
    def __init__(self, bars_to_play: Iterable[Bar]):
        super().__init__()
        self.bars_to_play = list(bars_to_play)

    def play(self, action: Callable[[Bar], None]):
        for bar in self.bars_to_play:
            self.bars.append(bar)
            self.reload_bar_series()
            action(bar)
        self.bars_to_play = []
